zeno/agents/kba/tool_kba_insights.py

```diff
@@ -0,0 +1,156 @@
+            )
+        return v
+
+    @validator("data")
+    def validate_data_type(cls, v, values):
+        insight_type = values.get("type")
+        if insight_type == InsightType.TEXT and not isinstance(v, str):
+            raise ValueError("Data must be a string for text insights")
+        elif insight_type == InsightType.TABLE and not isinstance(v, list):
+            raise ValueError(
+                "Data must be a list of dictionaries for table insights"
+            )
+        elif insight_type == InsightType.CHART and not isinstance(
+            v, (dict, ChartData)
+        ):
+            raise ValueError("Data must be ChartData for chart insights")
+        return v
+
+
+class InsightsResponse(BaseModel):
+    insights: List[Insight] = Field(
+        ...,
+        min_items=1,
+        description="List of insights generated from the analysis",
+    )
+
+
+class ColumnSelectionOutput(BaseModel):
+    columns: List[str] = Field(
+        ...,
+        description="List of column names relevant to the user query based on knowledge base, user persona, and user query",
+    )
+
+
+column_selection_agent = haiku.with_structured_output(ColumnSelectionOutput)
+insights_agent = haiku.with_structured_output(InsightsResponse)
+
+column_description = pd.read_csv("data/kba/kba_column_descriptions.csv")
+
+
+@tool("kba-insights-tool")
+def kba_insights_tool(question: str, state: Annotated[Dict, InjectedState]):
+    """Find insights relevant to the user query for the Key Biodiversity Areas (KBAs).
+
+    Args:
+        question: The user's question or query
+    """
+    print("kba insights tool")
+    kba_within_aoi = state["kba_within_aoi"]
+    kba_within_aoi = gpd.GeoDataFrame.from_features(json.loads(kba_within_aoi))
+    user_persona = state["user_persona"]
+
+    column_selection_prompt = KBA_COLUMN_SELECTION_PROMPT.format(
+        user_persona=user_persona,
+        question=question,
+        dataset_description=column_description.to_csv(index=False),
+    )
+    columns = column_selection_agent.invoke(
+        [
+            AIMessage(content=column_selection_prompt),
+            HumanMessage(content=question),
+        ]
+    ).columns
+    # add siteName and sitecode to the columns list if they are not already in the list
+    if "siteName" not in columns:
+        columns.append("siteName")
+    if "sitecode" not in columns:
+        columns.append("sitecode")
+    # remove geometry column if it is in the columns list
+    if "geometry" in columns:
+        columns.remove("geometry")
+
+    kba_within_aoi_filtered = kba_within_aoi[columns]
+
+    kba_insights_prompt = KBA_INSIGHTS_PROMPT.format(
+        user_persona=user_persona,
+        question=question,
+        column_description=column_description[
+            column_description["column"].isin(columns)
+        ].to_csv(index=False),
+        data=kba_within_aoi_filtered.to_csv(index=False),
+    )
+
+    response = insights_agent.invoke(kba_insights_prompt)
```

Comment:
Is there a way to ask for markdown output explicitly? Or is that already happening somewhere?

Reply:
Can be done by explicit prompting